storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -5486,59 +5486,141 @@ static std::string read_file_as_string(const std::string file) {
+
+  if (fil_space != NULL) {
+    char tmpname[FN_REFLEN];
+    char *oldpath, *space_name;
+    bool res =
+        fil_space_read_name_and_filepath(fil_space->id, &space_name, &oldpath);
+
+    if (!res || !os_file_exists(oldpath)) {
+      xb::error() << "prepare_handle_ren_files: Tablespace " << fil_space->name
+                  << " not found.";
+      ut::free(oldpath);
+      ut::free(space_name);
+      return false;
+    }
+
+    strncpy(tmpname, dest_space_name, sizeof(tmpname) - 1);
+
+    xb::info() << "prepare_handle_ren_files: renaming " << fil_space->name
+               << " to " << dest_space_name;
+
+    if (!fil_rename_tablespace(fil_space->id, oldpath, tmpname, NULL)) {
+      xb::error() << "prepare_handle_ren_files: Cannot rename "
+                  << fil_space->name << " to " << dest_space_name;
+      ut::free(oldpath);
+      ut::free(space_name);
+      return false;
+    }
+    // rename .delta .meta files as well
+    if (xtrabackup_incremental) {
+      std::string from_path =
+          entry.datadir + OS_PATH_SEPARATOR + std::string{space_name};
+      ;
+      std::string to_path = entry.datadir + ren_file_content;
+
+      rename_force(from_path + ".ibd.delta", to_path + ".delta");
+      rename_force(from_path + ".ibd.meta", to_path + ".meta");
+    }
+    // delete the .ren file, we don't need it anymore
+    os_file_delete(0, ren_path.c_str());
+    ut::free(oldpath);
+    ut::free(space_name);
+
+    return true;
   }
-  os_file_delete(0, ren_file.c_str());
-  return true;
+
+  xb::error() << "prepare_handle_ren_files(): failed to handle " << ren_path;
+  return false;
 }
-/* Handle DDL for deleted files */
+
+/**
+Handle DDL for deleted files
+example input: test/10.ibd.del file
+        -> tablespace with space_id = 10 will be deleted
+@return true on success */
 static bool prepare_handle_del_files(
     const datadir_entry_t &entry, /*!<in: datadir entry */
     void *arg __attribute__((unused))) {
   if (entry.is_empty_dir) return true;
 
+  std::string del_file_name = entry.file_name;
   std::string del_file = entry.path;
-  std::string dest_path = entry.rel_path;
-  xb::info() << "prepare_handle_del_files: " << del_file;
-#ifdef UNIV_DEBUG
-  size_t index = dest_path.find(".del");
-  assert(index != std::string::npos);
-#endif  // UNIV_DEBUG
-  os_file_delete(0, del_file.c_str());
-  del_file.erase(del_file.length() - 4);
-  delete_force(dest_path.substr(0, dest_path.length() - 4).c_str());
-  if (xtrabackup_incremental) {
-    delete_force(del_file + ".delta");
-    delete_force(del_file + ".meta");
-  }
+  space_id_t space_id;
 
-  return true;
+  // trim .ibd.del
+  space_id = atoi(del_file_name.substr(0, del_file_name.length() - 8).c_str());
+  fil_space_t *fil_space = fil_space_get(space_id);
+  if (fil_space != NULL) {
+    char *path, *space_name;
+    bool res =
+        fil_space_read_name_and_filepath(fil_space->id, &space_name, &path);
+
+    if (!res || !os_file_exists(path)) {
```

Comment:
Do we need to abort prepare if the file is already deleted? and path doesn't exist?